File: backend-nestjs/auth-service/src/menu/menu.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, IsNull, Or, Equal } from 'typeorm';
import { MenuEntity } from './menu.entity';

export interface MenuNode {
  id: number;
  name: string;
  code: string;
  menu_type: string;
  path: string | null;
  icon: string | null;
  permission_code: string | null;
  sort: number;
  is_visible: boolean;
  is_enabled: boolean;
  meta: Record<string, any> | null;
  children: MenuNode[];
}

@Injectable()
export class MenuService {
  constructor(
    @InjectRepository(MenuEntity)
    private readonly menuRepo: Repository<MenuEntity>,
  ) {}

  async getMenuTree(tenantId: number, permCodes: Set<string>): Promise<MenuNode[]> {
    const allMenus = await this.menuRepo.find({
      where: {
        isEnabled: true,
        tenantId: Or(IsNull(), Equal(tenantId)),
      },
      order: { sort: 'ASC' },
    });

    const visibleMenus = allMenus.filter(
      (m) => !m.permissionCode || permCodes.has(m.permissionCode),
    );

    return this.buildTree(visibleMenus);
  }

  private buildTree(menus: MenuEntity[]): MenuNode[] {
    const menuMap = new Map<number, MenuNode>();
    for (const m of menus) {
      menuMap.set(m.id, {
        id: m.id,
        name: m.name,
        code: m.code,
        menu_type: m.menuType,
        path: m.path,
        icon: m.icon,
        permission_code: m.permissionCode,
        sort: m.sort,
        is_visible: m.isVisible,
        is_enabled: m.isEnabled,
        meta: m.meta,
        children: [],
      });
    }

    const visibleIds = new Set(menuMap.keys());
    const roots: MenuNode[] = [];

    for (const m of menus) {
      const node = menuMap.get(m.id)!;
      if (m.parentId && visibleIds.has(m.parentId)) {
        menuMap.get(m.parentId)!.children.push(node);
      } else {
        roots.push(node);
      }
    }

    // 剪枝
    return roots.filter(
      (item) => item.menu_type !== 'directory' || item.children.length > 0,
    );
  }
}

File: backend-java/auth-service/src/main/java/com/cdm/auth/service/MenuService.java
package com.cdm.auth.service;

import com.cdm.auth.entity.MenuEntity;
import com.cdm.auth.repository.MenuRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
@RequiredArgsConstructor
public class MenuService {

    private final MenuRepository menuRepo;

    public List<Map<String, Object>> getMenuTree(Long tenantId, Set<String> permCodes) {
        var allMenus = menuRepo.findActiveMenus(tenantId);
        var visibleMenus = allMenus.stream()
                .filter(m -> m.getPermissionCode() == null
                        || m.getPermissionCode().isEmpty()
                        || permCodes.contains(m.getPermissionCode()))
                .toList();
        return buildTree(visibleMenus);
    }

    private List<Map<String, Object>> buildTree(List<MenuEntity> menus) {
        var menuMap = new LinkedHashMap<Long, Map<String, Object>>();
        for (var m : menus) {
            var node = new LinkedHashMap<String, Object>();
            node.put("id", m.getId());
            node.put("name", m.getName());
            node.put("code", m.getCode());
            node.put("menu_type", m.getMenuType());
            node.put("path", m.getPath());
            node.put("icon", m.getIcon());
            node.put("permission_code", m.getPermissionCode());
            node.put("sort", m.getSort());
            node.put("is_visible", m.getIsVisible());
            node.put("is_enabled", m.getIsEnabled());
            node.put("meta", m.getMeta());
            node.put("children", new ArrayList<Map<String, Object>>());
            menuMap.put(m.getId(), node);
        }

        var visibleIds = menuMap.keySet();
        var roots = new ArrayList<Map<String, Object>>();
        for (var m : menus) {
            var node = menuMap.get(m.getId());
            if (m.getParentId() != null && visibleIds.contains(m.getParentId())) {
                @SuppressWarnings("unchecked")
                var children = (List<Map<String, Object>>) menuMap.get(m.getParentId()).get("children");
                children.add(node);
            } else {
                roots.add(node);
            }
        }
        roots.removeIf(item -> "directory".equals(item.get("menu_type"))
                && ((List<?>) item.get("children")).isEmpty());
        return roots;
    }
}
